fault/fault.go

```diff
@@ -139,47 +139,62 @@ type streamInt struct {
 
 func (si *streamInt) RecvMsg(m any) error {
 	err := si.ServerStream.RecvMsg(m)
-	msg, err := si.int.sendRecvFault(si.fs.originMsgCh, si.rpcID, m, err)
+	msg, err := si.int.sendRecvFault(si.fs.originMsgCh, si.rpcID, m, faultpb.MessageType_MESSAGE_TYPE_REQUEST, err)
 	if pm, ok := m.(proto.Message); ok {
 		proto.Merge(pm, msg.(proto.Message))
 	}
-
 	return err
 }
 
 func (si *streamInt) SendMsg(m any) error {
-	msg, err := si.int.sendRecvFault(si.fs.originMsgCh, si.rpcID, m, nil)
+	msg, err := si.int.sendRecvFault(si.fs.originMsgCh, si.rpcID, m, faultpb.MessageType_MESSAGE_TYPE_RESPONSE, nil)
 	if err != nil {
 		return err
 	}
 	return si.ServerStream.SendMsg(msg)
 }
 
+// Stream implements the grpc strean server imterceptor interface, adding fault injection to streanubg RPCs.
 func (i *Interceptor) Stream(srv any, stream grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
 	if info.FullMethod == "/lemming.fault.FaultInject/Intercept" { // Do not self-intercept
 		return handler(srv, stream)
 	}
 
 	i.faultSubsMu.Lock()
-	var sub *faultSubscription
-	for _, fs := range i.faultSubs {
-		if fs.exp.FindString(info.FullMethod) != "" {
-			sub = fs
-			break
-		}
-	}
+	sub, ok := i.faultSubs[info.FullMethod]
 	i.faultSubsMu.Unlock()
+	if !ok {
+		return handler(srv, stream)
+	}
 
 	si := &streamInt{
 		ServerStream: stream,
 		int:          i,
 		fs:           sub,
 		rpcID:        uuid.New().String(),
 	}
-
-	return handler(srv, si)
+	hErr := handler(srv, si)
+	_, err := si.int.sendRecvFault(si.fs.originMsgCh, si.rpcID, nil, faultpb.MessageType_MESSAGE_TYPE_STREAM_END, hErr)
```

Comment:
Are we terminating the stream no matter what is returned from the handler?

Reply:
no, when the handler exits, we allow a final opportunity to inject a fault.